File: Convert Binary Search Tree to Doubly Linked List.cpp
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 * Definition of Doubly-ListNode
 * class DoublyListNode {
 * public:
 *     int val;
 *     DoublyListNode *next, *prev;
 *     DoublyListNode(int val) {
 *         this->val = val;
           this->prev = this->next = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param root: The root of tree
     * @return: the head of doubly list node
     */
    DoublyListNode* bstToDoublyList(TreeNode* root) {
        // Write your code here
        DoublyListNode dummy(INT_MIN);
        DoublyListNode *head = &dummy;
        doHelper(root, &head);
        //without the if-case, pass too. But, it should be remove the prev
        // if (dummy.next)
        // {
        //     dummy.next->prev = nullptr;
        // }
        return dummy.next;
    }
    
    void doHelper(TreeNode *root, DoublyListNode **head)
    {
        if (!root)
        {
            return;
        }
        doHelper(root->left, head);
        DoublyListNode *cur = new DoublyListNode(root->val);
        (*head)->next = cur;
        cur->prev = *head;
        *head = cur;
        doHelper(root->right, head);
        
    }
};

File: Convert Binary Search Tree to Doubly Linked List.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        this.val = val
        this.left, this.right = None, None
Definition of Doubly-ListNode
class DoublyListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = self.prev = next
"""

class Solution:
    """
    @param root, the root of tree
    @return: a doubly list node
    """
    def bstToDoublyList(self, root):
        # Write your code here
        
        def helper(treeNode):
            if not treeNode:
                return (None, None)
            
            midle = DoublyListNode(treeNode.val)
            minRet, maxRet = midle, midle
            (minLeft, maxLeft) = helper(treeNode.left)
            midle.pre = maxLeft
            if maxLeft:
                minRet, maxLeft.next = minLeft, midle
            (minRight, maxRight) = helper(treeNode.right)
            midle.next = minRight
            if minRight:
                maxRet, minRight.pre = maxRight, midle
            # minRet = midle if not minLeft else minLeft
            # maxRet = midle if not maxRight else maxRight
            return (minRet, maxRet)
        
        return helper(root)[0]
            